MaxText/sft/hooks.py

```diff
@@ -0,0 +1,177 @@
+#  limitations under the License.
+
+"""Training and data loading hooks for SFT"""
+
+from flax import nnx
+from collections import defaultdict
+import datetime
+import jax
+import jax.numpy as jnp
+
+from MaxText.metric_logger import MetricLogger
+from MaxText.utils import gcs_utils
+from MaxText import exceptions
+from MaxText import maxtext_utils
+from MaxText import max_logging
+from MaxText import max_utils
+from MaxText.data_loader import DataLoader
+from MaxText.input_pipeline.input_pipeline_interface import create_data_iterator
+from tunix.sft import peft_trainer
+from tunix.sft.hooks import DataHooks, TrainingHooks
+from typing_extensions import override
+from MaxText.utils.goodput_utils import (
+    GoodputEvent,
+    record_goodput,
+)
+
+
+class SFTTrainingHooks(TrainingHooks):
+  """Training hooks for SFT."""
+
+  def __init__(self, config, mesh, learning_rate_schedule, goodput_recorder):
+    self.config = config
+    self.mesh = mesh
+    self.metric_logger = MetricLogger(self.config, learning_rate_schedule)
+    self.goodput_recorder = goodput_recorder
+    self.metadata = {
+        "first_train_step": None,
+    }
+    self.train_metadata = defaultdict(float)
+    self.eval_metadata = defaultdict(float)
+
+  @override
+  def on_train_start(self, train_ctx: peft_trainer.PeftTrainer):
+    """Called at the beginning of training."""
+    state = nnx.state(train_ctx.model)
+    params = state.filter(nnx.Param)
+
+    if not self.config.using_pipeline_parallelism:
+      maxtext_utils.assert_params_sufficiently_sharded(params, self.mesh, self.config.sharding_tolerance)
+
+    self.metric_logger.write_setup_info_to_tensorboard(params)
+
+    if self.config.dump_hlo:
+      jax.block_until_ready(state)  # Ensure compilation has finished
+      gcs_utils.upload_dump(
+          self.config.dump_hlo_local_dir,
+          self.config.dump_hlo_gcs_dir,
+          module_name=self.config.dump_hlo_module_name,
+          delete_local_after=self.config.dump_hlo_delete_local_after,
+          all_host_upload=self.config.dump_hlo_upload_all,
+      )
+
+    self.metadata["first_train_step"] = train_ctx.train_steps
+
+  @override
+  def on_train_end(self, train_ctx: peft_trainer.PeftTrainer):  # pylint: disable=unused-argument
+    """Called at the end of training."""
+    self.metadata.clear()
+    self.train_metadata.clear()
+    self.eval_metadata.clear()
+    if self.metric_logger:
+      self.metric_logger.flush_metrics_and_cleanup()
+
+  @override
+  def on_train_step_start(self, train_ctx: peft_trainer.PeftTrainer):
+    """Called at the beginning of a training step."""
+    if self.config.enable_goodput_recording:
+      record_goodput(self.goodput_recorder, f"record_{GoodputEvent.STEP.value}_start_time", train_ctx.train_steps)
+
+    if train_ctx.data_hooks.train_batch is not None:
+      self.train_metadata["total_weights"] += jnp.sum(train_ctx.data_hooks.train_batch["targets_segmentation"] != 0)
```

Comment:
this is a tiny bit subtle (i initially wondered whether the != 0 was necessary, although i now see it obviously is) so maybe a comment is warranted